Tensile/KernelWriter.py

```diff
@@ -836,49 +1044,76 @@ def noLoadLoop( self, kernel, tensorParametersA, tensorParametersB, isOptNLL, pa
-            "7wait for local read"))
-      if self.enable["MAC"]:
-        luIdx = (u) % (self.numVgprBuffer+1) # local to use for MACs
-        pIdx = (u) % (self.numItersPLR+1) # local to use for MACs
-        if kernel["EnableMatrixInstruction"]:
-          kl.append(pack[pIdx])
-          for item in list(pack[pIdx].items()):
-            if item.tempVgpr != None:
-              self.vgprPool.checkIn(item.tempVgpr)
-              item.tempVgpr = None
-          kl.append(self.mfmaIter(kernel, luIdx, kernel["InnerUnroll"]))
-        else:
-          kl.append(self.macIter(kernel, luIdx, kernel["InnerUnroll"], useMacro=not isOptNLL ))
-    kl.append(self.closeSumAtLeastUnroll(kernel, prefetch=False, isOptNLL=isOptNLL))
+    # unrollI,J option adds iterations in noLoadLoop
+    # DepthU= mathces prolemsize[k] and no unroll Loop 
+    for UnrollI in range(0,kernel["NumITilesUnRoll"]):
+      for UnrollJ in range(0,kernel["NumJTilesUnRoll"]):
+        for u in range(0, kernel["LoopIters"]):
+          kl.append(self.comment("iter %u"%u))
+          plrIdx = (u+pflr) % (self.numVgprBuffer + 1)
+          plrPIdx = ((u+pflr) % (self.numItersPLR+1))
+          pack[plrPIdx] = Code.Module()
+          localReads = Code.Module()
+          waitCode = Code.Module()
+          macIterCode = Code.Module()
+          AccVgprReadCode = Code.Module("AccVgprRead")
+          for iui in range(0,kernel["InnerUnroll"]):
+            if self.enable["LocalRead"]:
+              if u < kernel["LoopIters"]/self.numIterPerCoalescedRead - self.numItersPLR:
+                if iui*self.numReadsIterCoalesced < kernel["InnerUnroll"]:
+                  # Reading 16-bit data from LDS requires packing when ECC enabled
+                  localReads.addText(self.comment("local read a"))
+                  localReadCodeA, packCodeA = self.localReadDo(kernel, plrIdx*self.numIterPerCoalescedRead, iui*self.numReadsIterCoalesced, 0, tensorParametersA)
+                  localReads.addCode(localReadCodeA)
+                  localReads.addText(self.comment("local read b"))
+                  localReadCodeB, packCodeB = self.localReadDo(kernel, plrIdx*self.numIterPerCoalescedRead, iui*self.numReadsIterCoalesced, 0, tensorParametersB)
+                  localReads.addCode(localReadCodeB)
+                  pack[plrPIdx].addCode(packCodeA)
+                  pack[plrPIdx].addCode(packCodeB)
+                  #if (kernel["NumITilesUnRoll"] + kernel["NumJTilesUnRoll"] > 2 and self.scheduleIterAlg ==2): 
+                  localReads.addText(self.comment("local read inc a"))
+                  localReads.addText(self.localReadInc(kernel, iui, tensorParametersA))
+                  localReads.addText(self.comment("local read inc b"))
+                  localReads.addText(self.localReadInc(kernel, iui, tensorParametersB))
+          # use only when unorllI,unrollJ != 1
+          if not (self.scheduleLastIterAlg):  
+            if localReads:
+              k1.append(localReads) 
```

Comment:
This section could be restructured: if using scheduleLastIterAlg, the loop doesn't need to run at all, and putting the check before the loop would also make it easier to read.